pkg/resource/cluster/hook.go

```diff
@@ -166,28 +166,34 @@ func (rm *resourceManager) customUpdate(
 	exit := rlog.Trace("rm.customUpdate")
 	defer exit(err)
 
+	// For asynchronous updates, latest(from ReadOne) contains the
+	// outdate values for Spec fields. However the status(Cluster status)
+	// is correct inside latest.
+	// So we construct the updatedRes object from the desired resource to
+	// obtain correct spec fields and then copy the status from latest.
+	updatedRes := rm.concreteResource(desired.DeepCopy())
 	if clusterDeleting(latest) {
 		msg := "Cluster is currently being deleted"
-		ackcondition.SetSynced(desired, corev1.ConditionFalse, &msg, nil)
-		return desired, requeueWaitWhileDeleting
+		// set the status from latest resource
+		updatedRes.SetStatus(latest)
+		ackcondition.SetSynced(updatedRes, corev1.ConditionFalse, &msg, nil)
+		return updatedRes, requeueWaitWhileDeleting
 	}
 	if !clusterActive(latest) {
 		msg := "Cluster is in '" + *latest.ko.Status.Status + "' status"
-		ackcondition.SetSynced(desired, corev1.ConditionFalse, &msg, nil)
+		// set the status from latest resource
+		updatedRes.SetStatus(latest)
+		ackcondition.SetSynced(updatedRes, corev1.ConditionFalse, &msg, nil)
 		if clusterHasTerminalStatus(latest) {
-			ackcondition.SetTerminal(desired, corev1.ConditionTrue, &msg, nil)
-			return desired, nil
+			ackcondition.SetTerminal(updatedRes, corev1.ConditionTrue, &msg, nil)
+			return updatedRes, nil
 		}
-		return desired, requeueWaitUntilCanModify(latest)
+		return updatedRes, requeueWaitUntilCanModify(latest)
 	}
 
-	// Merge in the information we read from the API call above to the copy of
-	// the original Kubernetes object we passed to the function
-	ko := desired.ko.DeepCopy()
-
 	// None of these methods modify the status, so we should return the latest
 	// status as given by the ReadOne
-	ko.Status = latest.ko.Status
+	updatedRes.SetStatus(latest)
```

Comment:
It seems like all of these conditions are using this, so let's move it under line 174?

Reply:
Good callout